metricbeat/module/docker/network/helper.go

```diff
@@ -64,18 +75,129 @@ type NetStats struct {
 	TxDropped     float64
 	TxErrors      float64
 	TxPackets     float64
+	Netstat       *sysinfotypes.NetworkCountersInfo
 	Total         *types.NetworkStats
 }
 
-func (n *NetService) getNetworkStatsPerContainer(rawStats []docker.Stat, dedot bool) []NetStats {
+// NoSumStats lists "stats", often config/state values, that can't be safely summed across PIDs
+var NoSumStats = []string{
+	"RtoAlgorithm",
+	"RtoMin",
+	"RtoMax",
+	"MaxConn",
+	"Forwarding",
+	"DefaultTTL",
+}
+
+func (n *NetService) getNetworkStatsPerContainer(client *client.Client, timeout time.Duration, rawStats []docker.Stat, cfg Config) ([]NetStats, error) {
 	formattedStats := []NetStats{}
 	for _, myStats := range rawStats {
+
+		var stats *sysinfotypes.NetworkCountersInfo
+		var err error
+		if cfg.NetworkSummary {
+			stats, err = fetchContainerNetStats(client, timeout, myStats.Container.ID)
+			if err != nil {
+				return nil, errors.Wrap(err, "error fetching per-PID stats")
+			}
+		}
+
 		for nameInterface, rawnNetStats := range myStats.Stats.Networks {
-			formattedStats = append(formattedStats, n.getNetworkStats(nameInterface, rawnNetStats, myStats, dedot))
+			singleStat := n.getNetworkStats(nameInterface, rawnNetStats, myStats, cfg.DeDot)
+			if cfg.NetworkSummary {
+				singleStat.Netstat = stats
+			}
+			formattedStats = append(formattedStats, singleStat)
 		}
 	}
 
-	return formattedStats
+	return formattedStats, nil
+}
+
+// fetchContainerNetStats gathers the PIDs associated with a container, and then uses go-sysinfo to grab the /proc/[pid]/net counters and sum them across PIDs.
+func fetchContainerNetStats(client *client.Client, timeout time.Duration, container string) (*sysinfotypes.NetworkCountersInfo, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	body, err := client.ContainerTop(ctx, container, []string{})
+	if err != nil {
+		return &sysinfotypes.NetworkCountersInfo{}, errors.Wrap(err, "error fetching container pids")
+	}
+
+	var pidPos int
+	for pos := range body.Titles {
+		if body.Titles[pos] == "PID" {
+			pidPos = pos
+			break
+		}
+	}
+
+	summedMetrics := &sysinfotypes.NetworkCountersInfo{}
+	for _, pid := range body.Processes {
+		strPID := pid[pidPos]
+		intPID, err := strconv.Atoi(strPID)
+		if err != nil {
+			return &sysinfotypes.NetworkCountersInfo{}, errors.Wrap(err, "error converting PID to int")
+		}
+
+		proc, err := sysinfo.Process(intPID)
+		procNet, ok := proc.(sysinfotypes.NetworkCounters)
+		if !ok {
+			break
+		}
+
+		counters, err := procNet.NetworkCounters()
+		if err != nil {
+			return &sysinfotypes.NetworkCountersInfo{}, errors.Wrapf(err, "error fetching network counters for PID %d", intPID)
+		}
+
+		summedMetrics = sumCounter(summedMetrics, counters)
```

Comment:
If I understand this correctly I wonder if we are summing here several times the same data. I think that all processes in the same network namespace report the same network counters.

For example if I start two listening netcats in a container, I see the two sockets under both pids:
```
$ for pid in $(pidof nc);do echo $pid; cat /proc/$pid/net/tcp; done 
3119046
  sl  local_address rem_address   st tx_queue rx_queue tr tm->when retrnsmt   uid  timeout inode                                                     
   0: 00000000:1F9B 00000000:0000 0A 00000000:00000000 00:00000000 00000000     0        0 36489370 1 0000000000000000 100 0 0 10 0                  
   1: 00000000:1F9C 00000000:0000 0A 00000000:00000000 00:00000000 00000000     0        0 36489816 1 0000000000000000 100 0 0 10 0                  
3118539
  sl  local_address rem_address   st tx_queue rx_queue tr tm->when retrnsmt   uid  timeout inode                                                     
   0: 00000000:1F9B 00000000:0000 0A 00000000:00000000 00:00000000 00000000     0        0 36489370 1 0000000000000000 100 0 0 10 0                  
   1: 00000000:1F9C 00000000:0000 0A 00000000:00000000 00:00000000 00000000     0        0 36489816 1 0000000000000000 100 0 0 10 0                  
``` 
I also see them if I start a cat inside the container:
```
$ for pid in $(pidof cat);do echo $pid; cat /proc/$pid/net/tcp; done 
3115294
  sl  local_address rem_address   st tx_queue rx_queue tr tm->when retrnsmt   uid  timeout inode                                                     
   0: 00000000:1F9B 00000000:0000 0A 00000000:00000000 00:00000000 00000000     0        0 36489370 1 0000000000000000 100 0 0 10 0                  
   1: 00000000:1F9C 00000000:0000 0A 00000000:00000000 00:00000000 00000000     0        0 36489816 1 0000000000000000 100 0 0 10 0                  
```
This also happens if I look to other files such as `/proc/$pid/net/netstat` or `/proc/$pid/net/snmp`, the content is the same for all the processes in the same namespace.

Not sure about the API call, but `docker container top` lists all the processes in the container, so here we may be summing several times the same information if we are looping over all the processes in the container.

I think that we would only need to report the counters of only one pid in the container. Or to be more correct, given that inside a container there can be processes of multiple network namespaces (dind case for example), for correctness we would need to sum the counters once per namespace found, that's it to sum the counters once per different `/proc/<pid>/ns/net` found.